### PR TITLE
Limit the number of screens displayed for signing review for Nano

### LINKED ISSUE #272: Show summary and warnings based on max_pages settings : Nano
- [ ]  Need to keep track of number of pages displayed
- [ ] check the max_pages settings and when pages exceed this values: ( Common for nano and stax)
- [ ] Parse the operations and calculate the summary of operations but do not display pages (Common for nano and stax)
- [ ] show   too many pages warning and then warning pages from Ledger ( for nano )
- [ ] show summary page (Nano)
- [ ] Allow the user to sign.

### PR DESCRIPTION
**Pull request by spalmer25**
Fixes #272 #270 #274 
 - [x] Add a flow for too long operation warning
 - [x] Add a flow to display the operation summary
 - [x] On `Allow Blindsigning For Large Tx` (default): display the too long warning flow and the operation summary flow
 - [x] On `Allow Blindsigning ON`: display the operation summary flow
 - [x] On `Allow Blindsigning OFF`: display normal

### CODE REVIEW

File: app/src/apdu_sign.c [hunk 1]
```diff
@@ -482,16 +483,78 @@ stream_cb(tz_ui_cb_type_t cb_type)
 }
 
 #ifdef HAVE_BAGL
+static void
+push_next_summary_screen(void)
+{
+#define FINAL_HASH global.keys.apdu.hash.final_hash
+
+    char num_buffer[TZ_DECIMAL_BUFFER_SIZE(TZ_NUM_BUFFER_SIZE / 8)] = {0};
```
Comment (ajinkyaraj-23) [inline app/src/apdu_sign.c:491]: Can we reuse the char buffer for printing numerical values global....parser_state->buffers.num.decimal as i did it in my PR.? As at this stage we have parsed everything and we have the final data with us. 
You can consider it for NANOS if the memory is falling short. 
Same for the hash_buffer.

File: app/src/apdu_sign.c [hunk 2]
```diff
@@ -482,16 +483,78 @@ stream_cb(tz_ui_cb_type_t cb_type)
...
+    char hash_buffer[TZ_BASE58_BUFFER_SIZE(sizeof(FINAL_HASH))]     = {0};
+    tz_operation_state *op
+        = &global.keys.apdu.sign.u.clear.parser_state.operation;
+    summarysign_step_t *step = &global.keys.apdu.sign.u.summary.step;
```
Comment (ajinkyaraj-23) [inline app/src/apdu_sign.c:495]: How about using a #define instead of creating a new pointer just like FINAL_HASH

File: app/src/apdu_sign.c [hunk 3]
```diff
@@ -499,6 +516,58 @@ init_summary_stream(void)
...
+}
+
+static void
+init_too_many_screens_stream(void)
+{
+    tz_ui_stream_init(too_many_screens_stream_cb);
+
+    tz_ui_stream_push_not_warning_trusted(
+#ifndef TARGET_NANOS
+        "This is operation", "is too long to be\ndisplay correctly."
```
[the remark this answers is not in the document]
Comment (spalmer25) [reply app/src/apdu_sign.c:617]: `Proceed to blindsign.` takes 21 char, but a line can only contain 19.

File: app/src/apdu_sign.c [hunk 4]
```diff
@@ -622,17 +690,23 @@ handle_data_apdu_clear(command_t *cmd)
...
         TZ_CHECK(refill());
-        if ((global.keys.apdu.sign.step == SIGN_ST_WAIT_USER_INPUT)
-            && (global.step != ST_SWAP_SIGN)) {
+        if (global.keys.apdu.sign.step == SIGN_ST_WAIT_USER_INPUT) {
             tz_ui_stream();
         }
+        break;
+    case ST_SWAP_SIGN:
+    case ST_SUMMARY_SIGN:
+        TZ_CHECK(refill_all());
```
[the remark this answers is not in the document]
Comment (spalmer25) [reply app/src/apdu_sign.c:836]: An error can occur during the parsing.
In that case, `SWAP` will just fail. But for `SUMMARY` a request for `BLINDSIGN` is asked to the user. An input from the user is therefore needed.
I will split the case `SWAP` and `SUMMARY` because, indeed, SWAP will not have users input. :+1:

File: app/src/apdu_sign.c [hunk 5]
```diff
@@ -250,9 +254,17 @@ refill_blo_im_full(void)
         }
     }
 
-    wrote = tz_ui_stream_push(TZ_UI_STREAM_CB_NOCB, st->field_info.field_name,
-                              global.line_buf, TZ_UI_LAYOUT_BN,
-                              TZ_UI_ICON_NONE);
+    if ((N_settings.blindsign_status == ST_BLINDSIGN_LARGE_TX)
+        && (SCREEN_DISPLAYED >= NB_MAX_SCREEN_ALLOWED)) {
```
[the remark this answers is not in the document]
Comment (spalmer25) [reply app/src/apdu_sign.c:259]: I'm not sure to understand.
Being in `refill_blo_im_done` means that the data has been fully parsed.
So, almost, no more screens regarding data will be pushed.
The only cases where more screens need to be added is when the parser has ended but its string-buffer still containing data that need to be display. In that case, the function `refill_blo_im_full` is called (https://github.com/trilitech/ledger-app-tezos-wallet/blob/main/app/src/apdu_sign.c#L240-L243)

File: app/src/apdu_sign.c (hunk 1 above)
Comment (spalmer25) [reply app/src/apdu_sign.c:491]: The two buffer `num_buffer` and `hash_buffer` are local to the function and would not affect the NanoS overall memory.

I'm not that sure about this information, so I've tested the two ways for edge cases, by increasing the `BUFF_LEN` value and run the two problematic tests : `tz2_sign_micheline` and `tz3_sign_micheline`, and the two ways act the same way (either both succeed or both fail).

I've used local buffers because it makes more sense than using a buffer for other purposes.

In any case, this function uses the part `summary` in the union `u` of `apdu_sign_state_t` it will not be reasonable/safe to use the part `clear` of this same union (even if it is working in practice)

File: app/src/apdu_sign.c (hunk 2 above)
Comment (spalmer25) [reply app/src/apdu_sign.c:495]: I'm not sure which is clearer. But I can do it !
Should I make a similar change for `blindsign_step_t *step = &global.keys.apdu.sign.u.blind.step;` from `bs_push_next(void)`?

File: app/src/apdu_sign.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (ajinkyaraj-23) [reply app/src/apdu_sign.c:620]: How about  "Accept Blindsign"

File: app/src/apdu_sign.c (hunk 2 above)
Comment (ajinkyaraj-23) [reply app/src/apdu_sign.c:498]: yeah the macro looks cleaner to me.

File: app/src/apdu_sign.c (hunk 1 above)
Comment (ajinkyaraj-23) [reply app/src/apdu_sign.c:494]: Sounds good. We have reduced memory in another MR so thats okay.

File: app/src/apdu_sign.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (ajinkyaraj-23) [reply app/src/apdu_sign.c:259]: No issues. I saw that we are calling iam_full from iam_done in that edge case. So thats alright.

File: app/src/apdu_sign.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (ajinkyaraj-23) [reply app/src/apdu_sign.c:620]: Use "Accept blindsign" for nanos and "proceed to \n blindsign" in nanosp and nanox.

**Review by ajinkyaraj-23 — APPROVED** [review]
LGTM
